### PR TITLE
[MERGE][IMP] mail, rating: delay rating request sending

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
FUNCTIONAL ISSUE

When they receive a rating request email, customers can click on this email to
pick a satisfaction level. However, this feedback is currently effectively
posted only when customers submit the feedback form.

Many customers do not go that far in the process and just close that form. This
means that while the feedback is caught and stored in DB nothing is posted and
users are not notified of this rating.

PURPOSE

When the customer clicks on an emoji, post a message in the chatter without
launching notifications. After 2 hours notify followers. If before this delay
the customer fills in the form, update the post content (new face, feedback)
and launch notification process with the feedback (inbox, emails, ...).

By default, the user has 2 hours to fill the feedback, after those 2 hours
we notify the message containing only the first rating without feedback.

TECHNICAL

Currently, rating from email (route with token) is the following

  * Customer receives an email;
  * Customer clicks one of the smiley face that redirects to a ``/rate`` route
    with the token and rate value;
  * the linked ``rating.rating`` record is updated and set consumed;
  * Customer may change its face and add a feedback. In that case a message
    is posted with the rating face in it;

Issues raised by this

  * first rating update does not notify anyone, no message is posted;
  * if Customer does not give a feedback, process is stuck at that point;
  * the rating is not really linked to the posted message (see ``message_id``
    field of ``rating`` record);

Fix that issue to have message directly at beginning of the flow and improve
log and notifications. Update ``rating_apply`` to post the message and update
it when feedback is given.

Allow to delay sending of notifications using a ``mail.message.schedule``
queue model that stores when to launch the notification process for a given
message.

See sub commits for more details.

LINKS

Task-2826699 (Mail: use datetime for scheduled_date mail field instead of char)
Task-2207626 (Rating: Log ratings, post feedbacks)

### CODE REVIEW

File: addons/test_mail/tests/test_mail_template.py [hunk 1]
```diff
@@ -61,6 +62,25 @@ def setUpClass(cls):
...
+        scheduled_datetime = datetime.datetime.now() + datetime.timedelta(days=3)
+
+        # schedule the mail in 3 days
+        self.test_template.scheduled_date = '{{datetime.datetime.now() + datetime.timedelta(days=3)}}'
+        mail_id = self.test_template.send_mail(self.test_record.id)
+        mail = self.env['mail.mail'].sudo().browse(mail_id)
+        self.assertEqual(
+            mail.scheduled_datetime.replace(second=0, microsecond=0),
+            scheduled_datetime.replace(second=0, microsecond=0),
+        )
```
Comment (xmo-odoo) [inline addons/test_mail/tests/test_mail_template.py:76]: Seems pretty racey, with no freezegun and `send_mail` being non-trivial it's eminently possible that the clock will go through a time boundary between the two, making this a nondetermininstic test.

Should either freeze time or use `assertAlmostEqual` with a small `delta=timeldeta(...)`.

File: addons/test_mail/tests/test_mail_template.py [hunk 2]
```diff
@@ -61,6 +62,25 @@ def setUpClass(cls):
...
+        self.assertEqual(mail.state, 'outgoing')
+
+        # check a wrong format
+        self.test_template.scheduled_date = '{{"test " * 5}}'
+        mail_id = self.test_template.send_mail(self.test_record.id)
+        mail = self.env['mail.mail'].sudo().browse(mail_id)
```
Comment (xmo-odoo) [inline addons/test_mail/tests/test_mail_template.py:82]: ???

File: addons/test_mail/tests/test_message_post.py [hunk 3]
```diff
@@ -540,6 +542,47 @@ def test_message_post_keep_emails(self):
...
+            )
+        self.assertFalse(self._new_mails)
+        self.assertFalse(self._mails)
+
+        messages_scheduled = self.env['mail.message.schedule'].search([('mail_message_id', '=', msg.id)])
+        self.assertEqual(len(messages_scheduled), 1, msg='Should have scheduled the message')
+        self.assertEqual(
+            messages_scheduled.scheduled_datetime.replace(second=0, microsecond=0),
+            scheduled_datetime,
+        )
```
Comment (xmo-odoo) [inline addons/test_mail/tests/test_message_post.py:569]: Probably same problem as above?

**Comment by tde-banana-odoo** [comment]
Is that a reminder that "you should create PRs in draft" :D ?

File: addons/mail/models/mail_template.py [hunk 4]
```diff
@@ -205,6 +206,15 @@ def generate_email(self, res_ids, fields):
                 values = results[res_id]
                 if values.get('body_html'):
                     values['body'] = tools.html_sanitize(values['body_html'])
+                if 'scheduled_date' in values:
+                    if values['scheduled_date']:
+                        try:
+                            values['scheduled_datetime'] = parse(values['scheduled_date'])
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/models/mail_template.py]: That could be funny. Odoo is trying to help !

File: addons/test_mail/tests/test_mail_template.py (hunk 1 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_template.py:76]: It was planned to use freezegun indeed :)

File: addons/mail/views/mail_menus.xml [hunk 5]
```diff
@@ -62,6 +62,11 @@
         parent="mail.mail_menu_technical"
         action="action_view_mail_message"
         sequence="1"/>
+    <menuitem name="Scheduled Messages"
+        id="mail_message_schedule_menu"
+        parent="mail.mail_menu_technical"
+        action="mail_message_schedule_action"
+        sequence="1"/>
```
Comment (dbeguin) [inline addons/mail/views/mail_menus.xml:69]: why same sequence ?

File: addons/mail/models/mail_thread.py [hunk 6]
```diff
@@ -2381,7 +2381,7 @@ def _notify_thread_by_inbox(self, message, recipients_data, msg_vals=False, **kw
         self.env['bus.bus'].sudo()._sendmany(bus_notifications)
```
Comment (dbeguin) [inline addons/mail/models/mail_thread.py:2381]: Commit message: It gives the user ~~has~~ some time to write their message.

File: addons/mail/models/mail_thread.py [hunk 7]
```diff
@@ -2436,7 +2441,12 @@ def _notify_thread_by_email(self, message, recipients_data, msg_vals=False,
 
         email_layout_xmlid = msg_vals.get('email_layout_xmlid') if msg_vals else message.email_layout_xmlid
         template_xmlid = email_layout_xmlid if email_layout_xmlid else 'mail.message_notification_email'
-        base_mail_values = self._notify_by_email_get_base_mail_values(message, additional_values={'auto_delete': mail_auto_delete})
+        base_mail_values = self._notify_by_email_get_base_mail_values(
+            message,
+            additional_values={'auto_delete': mail_auto_delete,
+                               'scheduled_date': scheduled_date,
+                              },
```
Comment (dbeguin) [inline addons/mail/models/mail_thread.py:2448]: beurky linter is beurky - here is my p'tite contribution

**Review by dbeguin — COMMENTED** [review]
kusjes

**Comment by tde-banana-odoo** [comment]
@dbeguin Done !

**Comment by tde-banana-odoo** [comment]
@robodoo override=ci/security

Reason: using Markup to avoid injecting stuff (feedback is a free text input customers can give and that is posted as body for rating message)

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

**Pull request closed without merging**